### ISSUE TITLE
`unused binding` reported if only one expression uses binding in reader conditional.

### ISSUE THREAD

**Issue by bhb** [borkdude/clj-kondo#282]
**version**

`v2019.06.17-alpha-SNAPSHOT`

**platform**

MacOS 10.13.6

**problem**

`unused binding` reported if only one expression uses binding in reader conditional.

```clojure
;; foo.cljc
(ns foo)

(defn foo [x]
  #?(:clj x
     :cljs 1))
```

`clj-kondo --lint foo.cljc`

**expected behavior**

No warning, since `x` is used in one expression.

**actual behavior**

`foo.cljc:3:12: warning: unused binding x`

**Comment by borkdude** [comment]
@bhb The way it currently works: `.cljc` is analyzed twice - once for clj and cljs. So for cljs the following would emit a warning in .cljc:

```
(defn bar []
  (future 1))
```

because `future` isn't a thing in CLJS.

This is why clj-kondo still reports that `x` is not used because that is true in CLJS. You can prevent that by using a different variable name in CLJS:

```
(defn foo [#?(:clj x
              :cljs _X)]
  #?(:clj x
     :cljs 1))
```

**Comment by borkdude** [comment]
@bhb Does the above make sense to you? I'm still wondering if it does, or if it should be changed to your expected behavior.

**Comment by bhb** [comment]
@borkdude Thanks for the explanation! I hadn’t considered the fact that some symbols like ‘future’ won’t be available in CLJS. I can certainly see why it makes sense to analyze the code twice, once for CLJ and once for CLJS. 

Your solution makes sense now that I understand how it works. I suspect it would add a fair bit of complexity to make it behave as I expected and really this is a corner case so I’m happy to use your workaround the few places this comes up. 

Thanks for making clj-kondo. It’s great!

**Comment by borkdude** [comment]
Thanks!

**Comment by vemv** [comment]
Hi!

I'm facing this very same issue. I understand the explanation and workaround, but that workaround doesn't seem sound to me.

Isn't that `#?(:clj x :cljs _x)` bending code (its comprehensibility) just to satisfy clj-kondo?

Generally clj-kondo makes a codebase cleaner, not the opposite.

So, wouldn't a sound logic rule be: if for any branch of `clj|cljs` a variable is used, then it's considered used for both branches? i.e. don't consider both branches completely independent worlds.

**Comment by borkdude** [comment]
@vemv Yes and no. It's often been helpful to me to detect that I didn't handle something in the `:cljs` side of things. So it's a double edged sword.

**Comment by vemv** [comment]
I didn't find it so bad after a few usages 👍

Maybe it's not code that anyone would ever write in absence of a linter, but the result is, in fact, informative.
